source/ShopPanel.h

```diff
@@ -35,6 +35,29 @@ class Ship;
 // Class representing the common elements of both the shipyard panel and the
 // outfitter panel (e.g. the sidebar with the ships you own).
 class ShopPanel : public Panel {
+public:
+
+	// The result of an attempt to buy. Implicitly converted from a string or
+	// boolean in code. Any string indicates failure. True indicates success,
+	// of course, while false indicate failure, but no need to pop up a message.
+	class BuyResult {
+	public:
+		BuyResult(const char* error) : success(false), message(error) {}
+		BuyResult(const std::string& error) : success(false), message(error) {}
+		BuyResult(bool result) : success(result), message() {}
+		BuyResult() = delete;
+
+		operator bool() const { return success; }
+		operator const std::string&() const { return message; }
+
+		bool Succeeded() const { return success;}
+		bool HasMessage() const { return message != ""; }
+		const std::string& Message() const { return message; }
+	private:
+		const bool success = true;
+		const std::string message;
```

Comment:
Yeah those const changes are probably good with the manual use of std::move. 

I'm a little confused at C++ right now, because adding `explicit` to the `operator bool` did nothing to the code that used it - I can still use `if(canBuy())` and it uses that operator, implicitly? Or is the fact it's in an if statement explicit enough for it? Fine.

I like the string operator for the non-verbose syntax of `new Dialog(result)` instead of `new Dialog(result.Message())` but that can change.

I found the `const char*` constructor was needed because it used the bool constructor instead of the string constructor - unless, again, the messages are made more verbose as `return string("error")` instead of `return "error"`

Reply:
> adding `explicit` to the `operator bool` did nothing to the code that used it - I can still use `if(canBuy())` and it uses that operator, implicitly? Or is the fact it's in an if statement explicit enough for it? Fine.

Yes, inside an `if` context is considered explicit for the boolean operator.

> I like the string operator for the non-verbose syntax of `new Dialog(result)` instead of `new Dialog(result.Message())` but that can change.

It's a no from me to _add a string conversion operator_ for the goal of not typing `.Message()` once.


> I found the `const char*` constructor was needed because it used the bool constructor instead of the string constructor - unless, again, the messages are made more verbose as `return string("error")` instead of `return "error"`

If you mark the one-argument constructors explicit (or at least the one that takes a boolean) then you shouldn't have an issue with the `const char *` argument constructing `BuyResult(true)`.

Reply:
> If you mark the one-argument constructors explicit (or at least the one that takes a boolean) then you shouldn't have an issue with the const char * argument constructing BuyResult(true).

Darn, I'm not finding this so easy. When I added `explicit` to the bool constructor:

`return false;` => `error: could not convert 'false' from 'bool' to 'const ShopPanel::BuyResult'`

`return {false};` => `error: converting to 'ShopPanel::BuyResult' from initializer list would use explicit constructor 'ShopPanel::BuyResult::BuyResult(bool)'`

> (You can likely avoid typing BuyResult(...) each time by wrapping the return value in curly braces, e.g. return { false };.)

So it seems that unlike the case where `explicit operator bool()` is used inside `if()`, a `return` isn't enough for the compiler to use `explicit` calls. 

> single-arg constructors are marked explicit unless the implicit conversion is guaranteed sensible.

So I'm leaning on keeping these constructors as-is since I think that implicit conversions are indeed guaranteed sensible - the point of BuyResult is that it's an easy stand-in for string or bool.